File: src/main/kotlin/com/advent2022/puzzle3/Puzzle3.kt
package com.advent2022.puzzle3

import com.advent2021.base.Base

typealias Data = ArrayList<String>
typealias Solution = Int
typealias Solution2 = Solution

// https://adventofcode.com/2022/day/3
fun main() {
    try {
        val puz = Puzzle3()
        val solution1 = puz.solvePuzzle("inputs.txt", Data())
        println("Solution1: $solution1")

        val solution2 = puz.solvePuzzle2("inputs.txt", Data())
        println("Solution2: $solution2")
    } catch (t: Throwable) {
        t.printStackTrace()
    }
}

class Puzzle3 : Base<Data, Solution?, Solution2?>() {
    override fun parseLine(line: String, data: Data) {
        data.add(line)
    }

    override fun computeSolution(data: Data): Solution {
        return data.sumOf { computeOne(it) }
    }

    private fun computeOne(items: String): Int {
        val string1 = items.substring(0, items.length / 2)
        val string2 = items.substring(items.length / 2)
        val set1 = string1.toSet()
        val set2 = string2.toSet()
        val common = set1.intersect(set2)
        return common.sumOf { charWorth(it) }
    }

    private fun charWorth(ch: Char): Int {
        val diff: Int =
            if (ch in 'a'..'z') {
                (ch - 'a' + 1)
            } else {
                (ch - 'A' + 27)
            }
        return diff
    }

    override fun computeSolution2(data: Data): Solution2 {
        val groups = data.chunked(3)
        return groups.sumOf { scoreGroup(it) }
    }

    private fun scoreGroup(group: List<String>): Int {
        val common = group.map { it.toSet() }.reduce { acc, g -> acc.intersect(g) }
        return common.sumOf { charWorth(it) }
    }
}

File: src/main/python/advent2024/Puzzle3.py
from Base import Base
import re

MyDataType = list[str]

class Puzzle(Base[MyDataType, int, int]):
    def main(self):
      solution1: int = self.solve_puzzle("puzzle3/inputs.txt", [])  # type: ignore
      print("Solution1: {}".format(solution1))

      solution2: int = self.solve_puzzle2("puzzle3/inputs.txt", []) # type: ignore
      print("Solution2: {}".format(solution2))

    def read_one_line(self, line: str, data: MyDataType):
      regex = re.compile("(mul\(\d+,\d+\)|do\(\)|don't\(\))")
      all_matches = regex.findall(line)
      data.extend(all_matches)

    def compute_solution(self, data: MyDataType) -> int:
      regex = re.compile("mul\((\d+),(\d+)\)")
      result = 0
      for element in data:
          match = regex.search(element)
          if match:
            result = result + (int(match.group(1)) * int(match.group(2)))
      return result

    def compute_solution2(self, data: MyDataType):
      regex = re.compile("mul\((\d+),(\d+)\)")
      result = 0
      do_flag = True
      for element in data:
          if element == "don't()":
              do_flag = False
          elif element == "do()":
              do_flag = True
          else:
              match = regex.search(element)
              if match and do_flag:
                result = result + (int(match.group(1)) * int(match.group(2)))
      return result
    
puzzle: Puzzle = Puzzle()
puzzle.main()